bundles/org.openhab.ui/web/src/assets/definitions/blockly/utils.js

```diff
@@ -13,31 +13,128 @@ export function addOSGiService (serviceName, serviceClass) {
-  const zdt = Blockly.JavaScript.provideFunction_(
+
+  let zdt = Blockly.JavaScript.provideFunction_(
     'zdt',
     ['var ' + Blockly.JavaScript.FUNCTION_NAME_PLACEHOLDER_ + ' = Java.type("java.time.ZonedDateTime");'])
-  const getZonedDateTime = Blockly.JavaScript.provideFunction_(
-    'getZonedDateTime',
-    [
-      'function ' + Blockly.JavaScript.FUNCTION_NAME_PLACEHOLDER_ + ' (datetime) {',
-      '  switch (datetime.length) {\n' +
-      `    case 10: return ${zdt}.parse(datetime + ' 00:00:00 +00:00', dtf.ofPattern('yyyy-MM-dd HH:mm:ss z'));\n` +
-      `    case 19: return ${zdt}.parse(datetime + ' +00:00', dtf.ofPattern('yyyy-MM-dd HH:mm:ss z'));\n` +
-      `    case 26: return ${zdt}.parse(datetime, dtf.ofPattern('yyyy-MM-dd HH:mm:ss z'));\n` +
+
+  let getzdt = Blockly.JavaScript.provideFunction_(
+    'getZonedDateTime', [
+      '/* Try to detect the format based on its length */',
+      'function ' + Blockly.JavaScript.FUNCTION_NAME_PLACEHOLDER_ + '(datetime) {',
+      '  datetime = datetime.replace(\'T\', \' \')',
+      '  switch (datetime.length) {',
+      `    case 10: return ${zdt}.parse(datetime + ' 00:00:00+00:00', dtf.ofPattern('yyyy-MM-dd HH:mm:ssz'));`,
+      `    case 16: return ${zdt}.parse(datetime + ':00+00:00', dtf.ofPattern('yyyy-MM-dd HH:mm:ssz'));`,
+      `    case 19: return ${zdt}.parse(datetime + '+00:00', dtf.ofPattern('yyyy-MM-dd HH:mm:ssz'));`,
+      `    case 25: return ${zdt}.parse(datetime, dtf.ofPattern('yyyy-MM-dd HH:mm:ssz'));`,
+      `    case 23: return ${zdt}.parse(datetime + ' +00:00', dtf.ofPattern('yyyy-MM-dd HH:mm:ss.SSS z'));`,
+      `    case 26: return ${zdt}.parse(datetime + ' +00:00', dtf.ofPattern('yyyy-MM-dd HH:mm:ss.SSSSSS z'));`,
+      `    case 29: return ${zdt}.parse(datetime, dtf.ofPattern('yyyy-MM-dd HH:mm:ss.SSSSz'));`,
+      `    case 32: return ${zdt}.parse(datetime, dtf.ofPattern('yyyy-MM-dd HH:mm:ss.SSSSSSSz'));`,
+      `    case 28: return ${zdt}.parse(datetime.slice(0,26) + ':' + datetime.slice(26,28), dtf.ofPattern('yyyy-MM-dd HH:mm:ss.SSSSz'));`,
+      `    default: return ${zdt}.parse(datetime);`,
+      '  }',
+      '}'
+    ])
+
+  let createzdt = Blockly.JavaScript.provideFunction_(
+    'createZonedDateTime', [
+      'function ' + Blockly.JavaScript.FUNCTION_NAME_PLACEHOLDER_ + '(year, month, day, hour, minute, second, nano, offsetString, timezoneString) {',
+      '  stringToParse = \'\' + year;',
+      '  stringToParse += \'-\' + (\'0\' + month).slice(-2);',
+      '  stringToParse += \'-\' + (\'0\' + day).slice(-2);',
+      '  stringToParse += \'T\' + (\'0\' + hour).slice(-2);',
+      '  stringToParse += \':\' + (\'0\' + minute).slice(-2);',
+      '  stringToParse += \':\' + (\'0\' + second).slice(-2);',
+      '  stringToParse += \'.\' + nano + offsetString + \'[\' + timezoneString + \']\';',
+      `  return ${zdt}.parse(stringToParse, dtf.ISO_ZONED_DATE_TIME);`,
```

Comment:
This seems to be an implementation of [`ZonedDateTime.of`](https://howtodoinjava.com/java/date-time/zoneddatetime-class/#22-create-zoneddatetime-with-values) so why not use it directly?

Since these functions are injected when you inject date support whether you need to call them of not, and are quite long, they tend to clutter users' scripts very quickly, so the fewer, the better. Besides we should IMHO use the best possible reference implementation when injecting things into user scripts.

Reply:
Good hint.

> so why not use it directly?

for two reasons:

1) I took that over from deibichs implementation directly
2) I wasn't aware of that zdt.of

I will see if I can optimize that eventually.